File: app/src/main/java/com/kongpf8848/dmail/MessageViewListFragment.java
package com.kongpf8848.dmail;

import android.app.Activity;
import android.os.Bundle;
import android.app.ListFragment;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import com.libmailcore.IMAPMessage;
import com.libmailcore.MailException;
import com.kongpf8848.dmail.mailcore.MailCore2Api;

public class MessageViewListFragment extends ListFragment {

    private static final String STATE_ACTIVATED_POSITION = "activated_position";

    private Callbacks mCallbacks = sDummyCallbacks;

    private int mActivatedPosition = ListView.INVALID_POSITION;

    private static final String TAG = "MessageViewListFragment";

    public interface Callbacks {
        /**
         * Callback for when an item has been selected.
         */
        void onItemSelected(IMAPMessage msg);
    }

    private static Callbacks sDummyCallbacks = new Callbacks() {
        @Override
        public void onItemSelected(IMAPMessage msg) {
        }
    };

    public MessageViewListFragment() {
    }

    private ArrayAdapter<MessageAdapter> adapter;
    private java.util.List<IMAPMessage> messages;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        MailCore2Api.getInstance().fetchMessages(Constants.INBOX, new UCallback<List<IMAPMessage>, MailException>() {
            @Override
            public void succeeded(List<IMAPMessage> imapMessages) {
                messages = imapMessages;
                updateResult();
            }

            @Override
            public void onFailed(MailException e) {

            }
        });

    }


    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (savedInstanceState != null && savedInstanceState.containsKey(STATE_ACTIVATED_POSITION)) {
            setActivatedPosition(savedInstanceState.getInt(STATE_ACTIVATED_POSITION));
        }
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        if (!(activity instanceof Callbacks)) {
            throw new IllegalStateException("Activity must implement fragment's callbacks.");
        }
        mCallbacks = (Callbacks) activity;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mCallbacks = sDummyCallbacks;
    }

    @Override
    public void onListItemClick(ListView listView, View view, int position, long id) {
        super.onListItemClick(listView, view, position, id);
        Log.d(TAG, "onListItemClick() called with: listView = [" + listView + "], view = [" + view + "], position = [" + position + "], id = [" + id + "]");
        mCallbacks.onItemSelected(messages.get((int) id));
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mActivatedPosition != ListView.INVALID_POSITION) {
            outState.putInt(STATE_ACTIVATED_POSITION, mActivatedPosition);
        }
    }


    private void setActivatedPosition(int position) {
        if (position == ListView.INVALID_POSITION) {
            getListView().setItemChecked(mActivatedPosition, false);
        } else {
            getListView().setItemChecked(position, true);
        }

        mActivatedPosition = position;
    }

    private void updateResult() {

        Log.d(TAG, "update result");
        ArrayList<MessageAdapter> array = new ArrayList();
        for(IMAPMessage msg : messages) {
            MessageAdapter msgAdapter = new MessageAdapter(msg);
            array.add(msgAdapter);
        }
        adapter = new ArrayAdapter<MessageAdapter>(
                getActivity(),
                android.R.layout.simple_list_item_activated_1,
                android.R.id.text1,
                array);

        setListAdapter(adapter);

    }
}

File: app/src/main/java/com/kongpf8848/dmail/messagelist/MessageViewListFragment.kt
package com.kongpf8848.dmail.messagelist

import com.libmailcore.IMAPMessage
import android.widget.ArrayAdapter
import android.os.Bundle
import com.kongpf8848.dmail.mailcore.MailCore2Api
import com.kongpf8848.dmail.util.UCallback
import com.libmailcore.MailException
import android.app.Activity
import android.app.ListFragment
import android.util.Log
import android.view.View
import android.widget.ListView
import com.kongpf8848.dmail.util.Constants
import java.util.ArrayList

class MessageViewListFragment : ListFragment() {
    private var mCallbacks = sDummyCallbacks
    private var mActivatedPosition = ListView.INVALID_POSITION

    interface Callbacks {
        /**
         * Callback for when an item has been selected.
         */
        fun onItemSelected(msg: IMAPMessage)
    }

    private var adapter: ArrayAdapter<MessageAdapter?>? = null
    private var messages: List<IMAPMessage>? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        MailCore2Api.instance.fetchMessages(
            Constants.INBOX,
            object : UCallback<List<IMAPMessage>, MailException> {
                override fun succeeded(imapMessages: List<IMAPMessage>) {
                    messages = imapMessages
                    updateResult()
                }

                override fun onFailed(e: MailException) {}
            })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        if (savedInstanceState != null && savedInstanceState.containsKey(STATE_ACTIVATED_POSITION)) {
            setActivatedPosition(savedInstanceState.getInt(STATE_ACTIVATED_POSITION))
        }
    }

    override fun onAttach(activity: Activity) {
        super.onAttach(activity)
        check(activity is Callbacks) { "Activity must implement fragment's callbacks." }
        mCallbacks = activity
    }

    override fun onDetach() {
        super.onDetach()
        mCallbacks = sDummyCallbacks
    }

    override fun onListItemClick(listView: ListView, view: View, position: Int, id: Long) {
        super.onListItemClick(listView, view, position, id)
        Log.d(
            TAG,
            "onListItemClick() called with: listView = [$listView], view = [$view], position = [$position], id = [$id]"
        )
        mCallbacks.onItemSelected(messages!![id.toInt()])
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        if (mActivatedPosition != ListView.INVALID_POSITION) {
            outState.putInt(STATE_ACTIVATED_POSITION, mActivatedPosition)
        }
    }

    private fun setActivatedPosition(position: Int) {
        if (position == ListView.INVALID_POSITION) {
            listView.setItemChecked(mActivatedPosition, false)
        } else {
            listView.setItemChecked(position, true)
        }
        mActivatedPosition = position
    }

    private fun updateResult() {
        Log.d(TAG, "update result")
        val array: ArrayList<MessageAdapter?> = ArrayList()
        for (msg in messages!!) {
            val msgAdapter = MessageAdapter(msg)
            array.add(msgAdapter)
        }
        adapter = ArrayAdapter(
            activity,
            android.R.layout.simple_list_item_activated_1,
            android.R.id.text1,
            array
        )
        listAdapter = adapter
    }

    companion object {
        private const val STATE_ACTIVATED_POSITION = "activated_position"
        private const val TAG = "MessageViewListFragment"
        private val sDummyCallbacks: Callbacks = object : Callbacks {
            override fun onItemSelected(msg: IMAPMessage) {}
        }
    }
}